sharding/mainchain/smc_client_test.go

```diff
@@ -1,6 +1,157 @@
+func (f *fakeClient) WaitForTransaction(ctx context.Context, hash common.Hash, durationInSeconds int64) error {
+
+	var receipt *types.Receipt
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Duration(durationInSeconds)*time.Second)
+
+	for err := error(nil); receipt == nil; receipt, err = f.backend.TransactionReceipt(ctxTimeout, hash) {
+
+		if err != nil {
+			cancel()
+			return fmt.Errorf("unable to retrieve transaction: %v", err)
+		}
+		if ctxTimeout.Err() != nil {
+			cancel()
+			return fmt.Errorf("transaction timed out, transaction was not able to be mined in the duration: %v", ctxTimeout.Err())
+		}
+	}
+	cancel()
+	ctxTimeout.Done()
+	log.Info(fmt.Sprintf("Transaction: %s has been mined", hash.Hex()))
+	return nil
+}
+
+// Creates and send Fake Transactions to the backend to be mined, takes in the context and
+// the current blocknumber as an argument and returns the signed transaction after it has been sent.
+func (f *fakeClient) CreateAndSendFakeTx(ctx context.Context, blocknumber int64) (*types.Transaction, error) {
+	tx := types.NewTransaction(uint64(blocknumber), common.HexToAddress("0x"), nil, 50000, nil, nil)
+	signedtx, err := types.SignTx(tx, types.MakeSigner(&params.ChainConfig{}, big.NewInt(blocknumber)), key)
+	if err != nil {
+		return nil, err
+	}
+	err = f.backend.SendTransaction(ctx, signedtx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send transaction: %v", err)
+	}
+	return signedtx, nil
+}
+
+func setup() *backends.SimulatedBackend {
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
+	return backend
+}
+
+// TestWaitForTransaction tests the WaitForTransaction function in the smcClient, however since for testing we do not have
+// an inmemory rpc server to interact with the simulated backend, we have to define the function in the test rather than
+// in `smc_client.go`.
+// TODO: Test the function in the main file instead of defining it here if a rpc server for testing can be
+// implemented.
+func TestWaitForTransaction(t *testing.T) {
+	backend := setup()
+	client := &fakeClient{backend: backend}
+	ctx := context.Background()
+	timeout := int64(5)
+
+	tx, err := client.CreateAndSendFakeTx(ctx, 0)
+	if err != nil {
+		t.Error(err)
+	}
+
+	receipt, err := client.backend.TransactionReceipt(ctx, tx.Hash())
+	if receipt != nil {
+		t.Errorf("transaction mined despite backend not being commited: %v", receipt)
+	}
+	err = client.WaitForTransaction(ctx, tx.Hash(), timeout)
+	if err == nil {
+		t.Error("transaction is supposed to timeout and return a error")
+	}
+
+	// Tests transaction timing out when the block is mined immediately
+	// in the timeout period
+	go func() {
+		newErr := client.WaitForTransaction(ctx, tx.Hash(), timeout)
+		if newErr != nil {
+			t.Errorf("transaction timing out despite backend being commited: %v", newErr)
+		}
+	}()
+	backend.Commit()
+	time.Sleep(time.Duration(timeout) * time.Second)
```

Comment:
Try using a channel to communicate when `WaitForTransaction` finishes instead

Reply:
using a timeout like this is brittle and increases the runtime of the test suite unnecessarily